File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-5.java
class Solution {
    public int characterReplacement(String s, int k) {
        int l = 0;
        int r = 1;
        int maxLen = 1;
        int maxFreq = 1;
        HashMap<Character, Integer> occ = new HashMap<>();
        occ.put(s.charAt(0), 1); 
        int sLen = s.length();

        while (r < sLen) {
            char atR = s.charAt(r);
            
            if (occ.containsKey(atR)) {
                int newOcc = occ.get(atR) + 1;
                occ.put(atR, newOcc);
                maxFreq = Math.max(maxFreq, newOcc);
            }
            else {
                occ.put(atR, 1);
            }

            if (r-l+1 <= maxFreq + k) {
                maxLen = Math.max(maxLen, r-l+1);
            }
            else {
                occ.put(s.charAt(l), occ.get(s.charAt(l))-1);
                l++;
            }

            r++;
        }

        return maxLen;
    }
}

File: Data Structures & Algorithms/leetcode/strings/is-anagram/submission-5.cpp
class Solution {
public:
    bool isAnagram(string s, string t) {
        vector<int> occ(26, 0);
        int l1 = s.size();
        int l2 = t.size();

        if (l1 != l2) {
            return false;
        }

        for (int i = 0; i < l1; i++) {
            occ[s[i]-'a'] += 1;
            occ[t[i]-'a'] -= 1;
        }

        for (int i = 0; i < 26; i++) {
            if (occ[i] != 0) {
                return false;
            }
        }
        
        return true;
    }
};
